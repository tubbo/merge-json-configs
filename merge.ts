import deepMerge from 'deepmerge'
import { jsonc } from 'jsonc'

/**
 * Merge multiple JSON configuration files together, with either one of them
 * optionally including comments. All comments and whitespace will be
 * stripped out, leaving just the merged data. Conflicting keys in
 * earlier files will be rewritten by the data in later files.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function mergeJsonConfigs<T = any>(...files: string[]) {
  return deepMerge.all<T>(
    files.map((file) => {
      try {
        return jsonc.readSync(file)
      } catch (_) {
        return {}
      }
    })
  )
}
