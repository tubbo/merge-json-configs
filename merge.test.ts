import { mergeJsonConfigs } from './merge'
import { jsonc } from 'jsonc'

interface TSConfig {
  compilerOptions: {
    target: string
    baseUrl: string
  }
  exclude: string[]
}

const ORIGINAL_FILE = '__mocks__/original.json'
const OVERRIDES_FILE = '__mocks__/overrides.json'

test('mergeJsonConfigs', async () => {
  const original = await jsonc.read(ORIGINAL_FILE)
  const overrides = await jsonc.read(OVERRIDES_FILE)
  const merged = mergeJsonConfigs<TSConfig>(ORIGINAL_FILE, OVERRIDES_FILE)

  expect(merged.compilerOptions.target).toEqual(original.compilerOptions.target)
  expect(merged.compilerOptions.baseUrl).toEqual(
    overrides.compilerOptions.baseUrl
  )
  expect(merged).toMatchSnapshot()
})
