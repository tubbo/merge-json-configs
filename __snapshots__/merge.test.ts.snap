// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mergeJsonConfigs 1`] = `
Object {
  "compilerOptions": Object {
    "baseUrl": "./",
    "declaration": true,
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "module": "commonjs",
    "outDir": "./dist",
    "rootDir": "./",
    "skipLibCheck": true,
    "sourceMap": true,
    "strict": true,
    "target": "es2016",
  },
  "exclude": Array [
    "generated.ts",
    "**/*.test.*",
    "**/*.spec.*",
    "dist/**/*",
  ],
}
`;
